apollo-router-core/src/query_planner/mod.rs

```diff
@@ -79,103 +78,120 @@ impl QueryPlan {
         service_registry: Arc<dyn ServiceRegistry>,
         schema: Arc<Schema>,
     ) -> Response {
-        let response = Arc::new(Mutex::new(Response::builder().build()));
         let root = Path::empty();
 
-        self.root
+        let (value, errors) = self
+            .root
             .execute_recursively(
-                Arc::clone(&response),
                 &root,
                 Arc::clone(&request),
                 Arc::clone(&service_registry),
                 Arc::clone(&schema),
+                &Value::default(),
             )
             .await;
 
-        // TODO: this is not great but there is no other way
-        Arc::try_unwrap(response)
-            .expect("todo: how to prove?")
-            .into_inner()
+        Response::builder().data(value).errors(errors).build()
     }
 }
 
 impl PlanNode {
     fn execute_recursively<'a>(
         &'a self,
-        response: Arc<Mutex<Response>>,
         current_dir: &'a Path,
         request: Arc<Request>,
         service_registry: Arc<dyn ServiceRegistry>,
         schema: Arc<Schema>,
```

Comment:
:yarn: I wonder if you actually need those arcs... :thinking: 
```suggestion
        request: &Request,
        service_registry: Arc<dyn ServiceRegistry>,
        schema: &Schema,
```

The one for the service registry could be replace by `&impl ServiceRegistry` I believe :thinking: 

:information_source: the good point about not using arcs in argument like that: you can pass a non-arc object or an arc one :brain: (so it's more flexible)

Reply:
Wait... could that also have perf impact? Because we're less incrementing the arf ref counter?

Reply:
for `Request` and `Schema`, done in 2c433c0, no issue. For `ServiceRegistry`, it's a bit trickier, with `&impl` it complains that the type is not `Sized` (well, yeah), I'll try by using `dyn` instead (since in most cases it will be a `dyn` anyway

Reply:
Have you tried adding `+ Sized` to the trait?

Reply:
for `ServiceRegistry`: a694c5e 
It looks fine so far, but it forces a trait object as argument

Reply:
looks cool